### PR TITLE
Listener: Add global conn limit opt out.

### PR DESCRIPTION
**Pull request by wez470**
Commit Message: Add listener global conn limit opt out.
Additional Description: Allows listeners (including the admin) to opt out of the globally set max downstream connections limit. Further info on why this is desired can be read in the linked issue. Individual listeners that have opted out of the global connection limit can still set a local limit using https://www.envoyproxy.io/docs/envoy/latest/configuration/listeners/runtime.
Risk Level: small
Testing: function, unit, integration
Docs Changes: Added api docs + mention in overload manager docs
Release Notes:
Fixes https://github.com/envoyproxy/envoy/issues/18678

### CODE REVIEW

File: api/envoy/config/listener/v3/listener.proto
Comment (adisuissa) [inline api/envoy/config/listener/v3/listener.proto:323]: I think an rst anchor next to the relevant sub-section in the doc (`docs/root/configuration/operations/overload_manager/overload_manager.rst`), and reference that here, just in case the comment will move somewhere else.
Same comment for the bootstrap.proto.

File: docs/root/configuration/operations/overload_manager/overload_manager.rst
Comment (adisuissa) [inline docs/root/configuration/operations/overload_manager/overload_manager.rst:158]: Probably should clarify that for the admin listener the config is not in the `Listener` proto, and point to the correct one.

**Review by adisuissa — COMMENTED** [review]
Thanks for working on this.
Left a couple of comments on API and docs.

**Comment by wez470** [comment]
@adisuissa PTAL, thanks.

File: envoy/event/dispatcher.h
Comment (adisuissa) [inline envoy/event/dispatcher.h:231]:
```suggestion
   * @param ignore_global_conn_limit controls whether the listener is limited by the global
```

File: envoy/event/dispatcher.h
Comment (adisuissa) [inline envoy/event/dispatcher.h:232]: Per style guide, comments should end with a period ('.').
(here and below)

File: envoy/network/listener.h
Comment (adisuissa) [inline envoy/network/listener.h:220]:
```suggestion
   * @return bool whether the listener should avoid blocking connections based on the globally set
```

File: envoy/server/configuration.h
Comment (adisuissa) [inline envoy/server/configuration.h:146]: Same as above.

**Review by adisuissa — COMMENTED** [review]
I just realized there are 3 things in play after this PR:
1. The global connections# limit (runtime flag)
2. The per-listener connections# limit (runtime flag)
3. The proposed ignore global-connections# limit field (for both the admin and general listeners).
It may be somewhat hard to understand which of these override the others.

Question: would it make sense to have just the global and local limits, where the local limit overrides the global, and if set to 0, it is unlimited?
I guess these can also be fields instead of runtime flags.

**Comment by wez470** [comment]
Thanks for the comments @adisuissa!

> Question: would it make sense to have just the global and local limits, where the local limit overrides the global, and if set to 0, it is unlimited?

This would satisfy the behaviour we want but I believe it's a question more so for maintainers. Making local limits override the global one is a breaking change. Today they can be used in tandem and a listener will be limited if it hits either limit. Further, AFAICT there's no way to set a local limit on the admin listener. So there's no way it could opt out without work to introduce some sort of local limit setting there.

**Comment by yanavlasov** [comment]
@wez470 would you be able to 

> Thanks for the comments @adisuissa!
> 
> > Question: would it make sense to have just the global and local limits, where the local limit overrides the global, and if set to 0, it is unlimited?
> 
> This would satisfy the behaviour we want but I believe it's a question more so for maintainers. Making local limits override the global one is a breaking change. Today they can be used in tandem and a listener will be limited if it hits either limit. Further, AFAICT there's no way to set a local limit on the admin listener. So there's no way it could opt out without work to introduce some sort of local limit setting there.

Listener connection limit allows overriding global limit down. Having a separate flag to override up makes sense to me. I think we could make it so that if the listener limit is larger than global limit it acts as override up, but it may just make config more complicated.

**Comment by wez470** [comment]
> Listener connection limit allows overriding global limit down. Having a separate flag to override up makes sense to me. I think we could make it so that if the listener limit is larger than global limit it acts as override up, but it may just make config more complicated.

I'll look into doing this. For the override up flag, were you thinking this would be a runtime flag?

**Comment by wez470** [comment]
I've realized that `Listener connection limit allows overriding global limit down. Having a separate flag to override up makes sense to me.` still misses some cases that the ignore global flag supports. If I want a listener that ignores the global limit, I have to make it handle more connections than the global limit. Although for my use case it would be okay to have a setup like this, there could be a case where you want a listener to only handle a handful of connections but still not be affected by the global limit. There would be no way to configure this with that approach. Thoughts?

File: source/common/network/tcp_listener_impl.h
Comment (yanavlasov) [inline source/common/network/tcp_listener_impl.h:46]: Can this be `const` ?

**Comment by KBaichoo** [comment]
Given the below diagram:


Listener Set (1)

I wonder if we'd want to end up with something more granular than global limits, but provide a grouping mechanism among subsets of listeners. e.g. Implement listener sets as in the middle of the diagram where listener 2 and 3 are limited by some maximum number, while listener 1 is not.

This would be a more generalized approach than global limit opt out, and enable cases such as multi-tenant connection limits.

I think it would add some additional configuration complexity (e.g. to set up groups), but the behavior seems more predictable / intuitive than overriding possibly occurring in both directions. Thoughts?

**Review by yanavlasov — CHANGES_REQUESTED** [review]
LGTM, modulo small const correctness nit.

/wait

**Comment by wez470** [comment]
@KBaichoo I like this idea. It is a little more work than I'd hoped to do to solve this issue personally, but I'm open to it if it's agreed that's the directions we should go.

**Review by KBaichoo — COMMENTED** [review]
@wez470 I'm also happy with the current solution (opt out) as it seems to be heading in the direction where we might eventually want to go and is more intuitive than overriding bidirectionally based on implicit values or implicit comparisons.

File: test/integration/cx_limit_integration_test.cc
Comment (wez470) [reply test/integration/cx_limit_integration_test.cc:282]: Nope, you're not missing anything. Sorry, I copied a previous test for these ones and modified to test what I needed. Oddly enough, that variable isn't being used in the test I copied from either. Removing from all 3 tests :)

File: docs/root/configuration/operations/overload_manager/overload_manager.rst
Comment (wez470) [reply docs/root/configuration/operations/overload_manager/overload_manager.rst]: I've added a sentence that mentions this. Please unresolve if you'd like to adjust wording.

**Comment by KBaichoo** [comment]
You might want to merge again since https://github.com/envoyproxy/envoy/pull/19007 disabled a flaky QUIC test.

**Comment by wez470** [comment]
Thanks

**Review by adisuissa — COMMENTED** [review]
/lgtm api

**Comment by wez470** [comment]
@yanavlasov Can you PTAL?

**Review by yanavlasov — APPROVED** [review]

**Comment by wez470** [comment]
Thanks all for reviewing!

**Pull request merged**
